plugins/out_stackdriver/stackdriver.c

```diff
@@ -277,6 +277,139 @@ static char *get_google_token(struct flb_stackdriver *ctx)
     return ctx->o->access_token;
 }
 
+static int process_local_resource_id(const void *data, size_t bytes,
+                                     struct flb_stackdriver *ctx, char *type)
+{
+    int i;
+    int len;
+    int ret = 0;
+    int flag = 0;
+    size_t off = 0;
+    msgpack_object k;
+    msgpack_object v;
+    msgpack_object root;
+    msgpack_object_map map;
+    msgpack_unpacked result;
+    char *local_resource_id;
+    char *ptr;
+    const char delim[] = ".";
+
+    msgpack_unpacked_init(&result);
+    while (msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS && !flag) {
```

Comment:
I get the logic here. However this looks complicated to me: why do we want to have a while follow by for?  

The logic in my mind should be something like this: 
1. check it is a valid msgpack for a record, return error if not. 
2. extract the `LOCAL_RESOURCE_ID_KEY` field out of it. There should be a common function for "extracting a string field from a map", return error if there is no such a field.
3. parse the right fields from the `local_resource_id`, return error during parsing. 

So overall what we want is: 1) reduce the nesting as much as possible. 2) return early as much as possible whenever an error happens.

Reply:
logic updated